versioned_docs/version-2020fa/lecture9.md

```diff
@@ -3,4 +3,224 @@ id: lecture9
+
+While it is safer to store hashed passwords, it is very easy to do it incorrectly, which is why we teach firebase authentication!
+
+### 2. Input Sanitization
+
+Input sanitization 'cleanses' inputs so that they can not be used in unintended ways.
+
+For example, it may seem reasonable to add every comment on a blog into a database entry like this:
+`“INSERT INTO 'comments' (comment) VALUES '” + user_input + “';”`, but it is actually incredibly unsafe
+if `user_input` were something like `'; DROP TABLE comments;`.
+
+Luckily for us, React DOM escapes values embedded in JSX before rendering them by default, preventing injection attacks.
+
+### 3. Don't Rely on Client Side Verification
+
+In general you cannot assume your endpoint will only be used by your frontend.
+
+Recall that we used Postman at the beginning of the semester to test Express endpoints. As such, it is always important to verify your inputs and authenticate your requests!
+
+### 4. Separation of Privileges
+
+Going hand-in-hand with the last point, when a request is made to your backend, make sure that the source of your request has the sufficient permissions to perform whatever action is being asked!
+
+For example, if the OfficeHours app recieved a request to edit the times to someone's office hours, the backend would check that this person is indeed the TA who owns that OH and not a student.
+
+## Deployment
+
+To deploy your web application means to put it on a Web server so others can access it via the internet. We will deploy frontend on Firebase and backend on Heroku.
+
+### Frontend Deployment
+
+To deploy to Firebase enter the following commands into terminal:
+
+```bash
+yarn global add firebase-tools
+yarn build
+firebase login
+firebase init
+<answer the questions>
+firebase deploy
+```
+
+`yarn build` will create a `build` directory containing a production build of your application.
+
+`firebase login` will prompt you to log in by opening up a web browser if you're not already signed it.
+
+`firebase init` will ask you the following questions:
+
+1. Which Firebase CLI features do you want to set up for this folder? Select **Hosting**.
+2. Associate with a Firebase project. **Select your Firebase project**
+3. What do you want as your public directory? **build**
+4. Configure as a single-page app (rewrite all urls to /index.html)? **Yes**
+5. Overwrite `index.html`? **No**
+
+Running `firebase deploy` will push your build assets to Firebase remote server and give you a URL to your live Firebase app site! Now you can share this site and access it over the internet.
+
+### Backend Deployment
+
+We will deploy backend on Heroku because deploying on Firebase is much more involved
+
+```bash
+yarn global add heroku
+git init
+git add .
+git commit -m "COMMIT MESSAGE"
+heroku login
+heroku create <optional project name>
+git push heroku master
+(optional) heroku open
+```
+
+You can then use the url generated as your backend endpoint in your frontend code.
```

Comment:
The slides no longer use global install. Follow the slides. In general, the entire deployment section needs to be rewritten to follow the slides!